### ISSUE TITLE
MQTT will_msg isn't set

### ISSUE THREAD

**Issue by UnknownProgrammer** [Ai-Thinker-Open/GPRS_C_SDK#203]
## 1. SDK version（SDK 版本）

V2.1.16
CSDTK 4.2

---

## 2. In what kind of operation problems appear, and how to reproduce the problem ?（什么样的操作步骤问题会出现，是否是稳定复现，如何复现问题？）

Using the MQTT last will functionality doesn't work entirely.

1. Adding will handling to the MQTT SSL Demo code:
```
void SecondTask(void *pData)
{
    MQTT_Event_t* event=NULL;
    semMqttStart = OS_CreateSemaphore(0);
    OS_WaitForSemaphore(semMqttStart,OS_WAIT_FOREVER);
    OS_DeleteSemaphore(semMqttStart);
    Trace(1,"start mqtt test");
    MQTT_Client_t* client = MQTT_ClientNew();
    MQTT_Connect_Info_t ci;
    MQTT_Error_t err;
    memset(&ci,0,sizeof(MQTT_Connect_Info_t));
    ci.client_id = CLIENT_ID;
    ci.client_user = CLIENT_USER;
    ci.client_pass = CLIENT_PASS;
    ci.keep_alive = 60;
    ci.clean_session = 1;
    ci.use_ssl = true;
    ci.ssl_verify_mode = MQTT_SSL_VERIFY_MODE_REQUIRED;
    ci.ca_cert = ca_crt;
    ci.ca_crl = NULL;
    ci.client_cert = client_crt;
    ci.client_key  = client_key;
    ci.client_key_passwd = NULL;
    ci.broker_hostname = BROKER_HOSTNAME;
    ci.ssl_min_version   = MQTT_SSL_VERSION_SSLv3;
    ci.ssl_max_version   = MQTT_SSL_VERSION_TLSv1_2;
    ci.entropy_custom    = "GPRS_A9";
    
    // Added will handling
    ci.will_topic = "thisisthewilltopic";
    ci.will_msg = "thisisthelastwillmsg";
    ci.will_qos = 2;
    ci.will_retain = true;
    
    err = MQTT_Connect(client,BROKER_IP,BROKER_PORT,OnMqttConnection,NULL,&ci);
    if(err != MQTT_ERROR_NONE)
        Trace(1,"MQTT connect fail,error code:%d",err);
    
    while(1)
    {
        if(OS_WaitEvent(secondTaskHandle, (void**)&event, OS_TIME_OUT_WAIT_FOREVER))
        {
            SecondTaskEventDispatch(event);
            OS_Free(event);
        }
    }
}
```

2. Connect to MQTT Server (Mosquitto 1.5)

Mosquiito log shows:

    1535232834: Will message specified (1 bytes) (r1, q2).
    1535232834: 	thisisthewilltopic

`will_topic`, `will_qos` and `will_retain` are set correctly, but `will_msg` is only recognized as one byte and therefore produces a empty output in will topic. This behaviour is independent from the length of the message. A short msg like "will" shows the same result.

The log output should look like this (will_msg = "thisisthelastwillmsg"):
```
1535296330: Will message specified (20 bytes) (r1, q2).
1535296330: 	thisisthewilltopic
```
---

**Comment by Neutree** [comment]
this error appears in mqtt ssl, but normal in mqtt without ssl, i'll fix it tomorrow

**Comment by Neutree** [comment]
Now fixed up in V2.1.18

**Comment by UnknownProgrammer** [comment]
I can confirm, problem solved with V2.1.18. Thank you.

**Comment by Axliyor** [comment]
Hi, I am unable to connect A9G to HiveMQ MQQT service!

MQTT connection status:5
MQTT connect to broker fail,error code:5
MQTT OnMqttConnection() end

I entered the following exactly, it connects on MQTT.FX but I can't connect via A9G

#define BROKER_IP  "b9dbb855b2d74ce5saad7c51e58f293b6.s2.eu.hivemq.cloud"//"mqtt.neucrack.com"
#define BROKER_PORT 8883
#define BROKER_HOSTNAME "b9dbb855b2d74ce5aad7c51e58f293b6.s2.eu.hivemq.cloud"
#define CLIENT_USER "AxliyorS"//"mqtt"
#define CLIENT_PASS "my_password"//"mqtt"
#define SUBSCRIBE_TOPIC "JHWFF89R28/app"
#define PUBLISH_TOPIC   "JHWFF89R28/gprs"

Any info would be appreciated! This is very important to me
